data_validation/schema_validation.py

```diff
@@ -216,40 +228,99 @@ def schema_validation_matching(
+    """Split the allow list string into a list of datatype:datatype tuples."""
+    # I've not moved this patter to a compiled constant because it should only
+    # happen once per command and I felt splitting the pattern into variables
+    # aided readability.
+    precision_scale_pattern = r"(?:\((?:[0-9 ,\-]+|'UTC')\))?"
+    nullable_pattern = r"(?:\[[noulabe\-]+\])?"
+    data_type_pattern = r"[a-z0-9 ]+" + precision_scale_pattern + nullable_pattern
+    csv_split_pattern = data_type_pattern + r":" + data_type_pattern
+    data_type_pairs = [
+        _.replace(" ", "").split(":")
+        for _ in re.findall(csv_split_pattern, allow_list_str, re.I)
+    ]
+    invalid_pairs = [_ for _ in data_type_pairs if len(_) != 2]
+    if invalid_pairs:
+        raise exceptions.SchemaValidationException(
+            f"Invalid data type pairs: {invalid_pairs}"
+        )
+    return data_type_pairs
+
+
+def expand_precision_range(s: str) -> list[str]:
+    """Expand an integer range (e.g. "0-3") to a list (e.g. ["0", "1", "2", "3"])."""
+    m_range = DECIMAL_PRECISION_SCALE_RANGE_PATTERN.match(s)
+    if not m_range:
+        return [s]
+    try:
+        p_lower = int(m_range.group(1))
+        p_upper = int(m_range.group(2))
+        if p_lower >= p_upper:
+            raise exceptions.SchemaValidationException(
+                f"Invalid allow list data type precision/scale: Lower value {p_lower} >= upper value {p_upper}"
+            )
+        return [str(_) for _ in range(p_lower, p_upper + 1)]
+    except ValueError as e:
+        raise exceptions.SchemaValidationException(
+            f"Invalid allow list data type precision/scale: {s}"
+        ) from e
 
 
-def get_datatype_name(st):
-    chars = []
-    for i in range(len(st)):
-        if ord(st[i].lower()) >= 97 and ord(st[i].lower()) <= 122:
-            chars.append(st[i].lower())
-    out = "".join(chars)
-    if out == "":
-        return -1
-    return out
+def expand_precision_or_scale_range(data_type: str) -> list[str]:
+    """Take a data type and example any precision/scale range.
+
+    For example "decimal(1-3,0)" becomes:
+      ["decimal(1,0)", "decimal(2,0)", "decimal(3,0)"]"""
+
+    m = DECIMAL_PRECISION_SCALE_PATTERN.match(data_type.replace(" ", ""))
+    if not m:
+        return [data_type]
+
+    if len(m.groups()) != 3:
+        raise exceptions.SchemaValidationException(
+            f"Badly formatted data type: {data_type}"
+        )
+
+    type_name, p, s = m.groups()
+    p_list = expand_precision_range(p)
+    if s:
+        s_list = expand_precision_range(s)
+        return_list = [
+            f"{type_name}({p},{s})" for p, s in itertools.product(p_list, s_list)
+        ]
+    else:
+        return_list = [f"{type_name}({_})" for _ in p_list]
+    return return_list
+
+
+def parse_allow_list(st: str) -> dict:
+    """Convert allow-list data type pairs into a dictionary like {key[value1, value2, etc], }"""
```

Comment:
what is the value in the dictionary here?

Reply:
It is how it was in the version of the code before I modified it:
```
# Converting allow list from string to map, to ease validation.
def parse_allow_list(st):
    output = {}
    stack = []
    key = None
    for i in range(len(st)):
        if st[i] == ":":
            key = "".join(stack)
            if key not in output:
                output[key] = []
            stack = []
            continue
        if st[i] == "," and not st[i + 1].isdigit():
            value = "".join(stack)
            output[key].append(value)
            stack = []
            i += 1
            continue
        stack.append(st[i])
    value = "".join(stack)
    output[key] = value
    stack = []
    return output
```

The value is that we can reference the list of valid target types for a source type directly without search for it. You can see this in action in `schema_validation_matching()`.